ocaml/src/alba_osd.ml

```diff
@@ -0,0 +1,281 @@
+              | Osd.Assert.Value (key, None) ->
+                 Lwt.return (Nsm_model.Assert.ObjectDoesNotExist (Slice.get_string_unsafe key))
+              | Osd.Assert.Value (key, Some blob) ->
+                 (* TODO can be optimized to not download the object
+                  * if the object was stored with a checksum *)
+                 let object_name = Slice.get_string_unsafe key in
+                 alba_client # download_object_to_bytes
+                             ~namespace
+                             ~object_name
+                             ~consistent_read:true
+                             ~should_cache:true >>= function
+                 | None -> Lwt.fail Nsm_model.Err.(Nsm_exn (Assert_failed, object_name))
+                 | Some (v, mf) ->
+                    if Osd.Blob.equal blob (Osd.Blob.Lwt_bytes v)
+                    then Lwt.return (Nsm_model.Assert.ObjectHasId
+                                       (object_name,
+                                        mf.Nsm_model.Manifest.object_id))
+                    else Lwt.fail Nsm_model.Err.(Nsm_exn (Assert_failed, object_name)))
+            asserts
+        in
+
+        Lwt.catch
+          (fun () ->
+           let nsm_asserts_t = prepare_asserts () in
+           let nsm_updates_t =
+             alba_client # nsm_host_access # with_namespace_id
+                         ~namespace
+                         prepare_updates
+           in
+           nsm_asserts_t >>= fun nsm_asserts ->
+           nsm_updates_t >>= fun nsm_updates ->
+
+           alba_client # nsm_host_access
+                       # with_nsm_client
+                       ~namespace
+                       (fun nsm ->
+                        nsm # apply_sequence nsm_asserts nsm_updates) >>= fun () ->
+
+           (* update the manifest cache *)
+           alba_client # nsm_host_access # with_namespace_id
+                       ~namespace
+                       (fun namespace_id ->
+                        List.iter
+                          (function
+                            | Nsm_model.Update.DeleteObject _ -> ()
+                            | Nsm_model.Update.PutObject (mf, _) ->
+                               Manifest_cache.ManifestCache.add
+                                 (alba_client # get_manifest_cache)
+                                 namespace_id mf.Nsm_model.Manifest.name mf)
+                          nsm_updates;
+                        Lwt.return ())
+           >>= fun () ->
+           Lwt.return Osd.Ok)
+          (function
+            | Nsm_model.Err.Nsm_exn (Nsm_model.Err.Assert_failed, key) ->
+               Lwt.return (Osd.Exn Osd.Error.(Assert_failed key))
+            | exn ->
+               Lwt.fail exn)
+    end
+  in
+  object(self :# Osd.osd)
+
+    method global_kvs =
+      get_kvs ~consistent_read:true prefix
+
+    method namespace_kvs namespace_id =
+      get_kvs ~consistent_read:false (to_namespace_name namespace_id)
```

Comment:
I was referring to the name of the function `to_namespace_name` which confusing.